src/shims/time.rs

```diff
@@ -1,8 +1,68 @@
+use std::sync::atomic::AtomicU64;
+use std::sync::Arc;
 use std::time::{Duration, Instant, SystemTime};
 
+use rustc_data_structures::sync::Ordering;
+
 use crate::concurrency::thread::Time;
 use crate::*;
 
+#[derive(Debug)]
+pub enum TimeAnchor {
+    Host(Instant),
+    Virtual(Arc<AtomicU64>),
+}
+
+impl TimeAnchor {
+    pub fn new(communicate: bool) -> Self {
+        if communicate { Self::Host(Instant::now()) } else { Self::Virtual(Arc::new(0.into())) }
+    }
+
+    pub fn get(&self) -> Duration {
+        match self {
+            Self::Host(instant) => Instant::now().saturating_duration_since(*instant),
+            Self::Virtual(nanoseconds) => Duration::from_nanos(nanoseconds.load(Ordering::Relaxed)),
+        }
+    }
+
+    pub fn tick(&self) {
+        match self {
+            Self::Host(_) => (),
+            Self::Virtual(nanoseconds) => {
+                nanoseconds.fetch_add(1, Ordering::Relaxed);
+            }
+        }
+    }
+
+    pub fn sleep(&self, duration: Duration) {
+        match self {
+            Self::Host(_) => std::thread::sleep(duration),
+            Self::Virtual(nanoseconds) => {
+                nanoseconds.fetch_add(duration.as_nanos().try_into().unwrap(), Ordering::Relaxed);
+            }
+        }
+    }
+
+    pub fn checked_add_since_now(&self, duration: Duration) -> Option<Time> {
+        match self {
+            Self::Host(_) => Instant::now().checked_add(duration).map(Time::Monotonic),
+            Self::Virtual(nanoseconds) =>
+                nanoseconds
+                    .load(Ordering::Relaxed)
+                    .checked_add(duration.as_nanos().try_into().unwrap())
+                    .map(|duration| Time::Virtual(duration, nanoseconds.clone())),
```

Comment:
why does this clone the `Arc` instead of taking a snapshot? Is that time meant to be stored somewhere across steps?

Reply:
from what I can see we'll want a snapshot here instead of the updating time, all other `Time` things are snapshots after all.

Reply:
I think this confusion is due to poor naming from my part. To know how much time remains before `Time` happens you need both the instant when time is due (first field) and the current instant (second field).

Reply:
oh 🤦 the other enum variants can just read the global time, but here we need the "global virtual time".

Reply:
document this in the enum please, that should avoid confusion in the future